tiatoolbox/wsicore/wsireader.py

```diff
@@ -2389,6 +2314,207 @@ def read_bounds(
+        im_region = np.array(im_region)
+
+        # Apply padding outside of the slide area
+        level_read_bounds = utils.transforms.locsize2bounds(
+            level_location, level_read_size
+        )
+        im_region = utils.image.crop_and_pad_edges(
+            bounds=level_read_bounds,
+            max_dimensions=level_size,
+            region=im_region,
+            pad_mode=pad_mode,
+            pad_constant_values=pad_constant_values,
+        )
+
+        # Resize to correct scale if required
+        im_region = utils.transforms.imresize(
+            img=im_region,
+            scale_factor=post_read_scale,
+            output_size=tuple(np.array(size).astype(int)),
+            interpolation=interpolation,
+        )
+
+        return utils.transforms.background_composite(image=im_region)
+
+    def read_bounds(
+        self,
+        bounds,
+        resolution=0,
+        units="level",
+        interpolation="optimise",
+        pad_mode="constant",
+        pad_constant_values=0,
+        coord_space="baseline",
+        **kwargs,
+    ):
+        # convert from requested to `baseline`
+        bounds_at_baseline = bounds
+        if coord_space == "resolution":
+            bounds_at_baseline = self._bounds_at_resolution_to_baseline(
+                bounds, resolution, units
+            )
+            _, size_at_requested = utils.transforms.bounds2locsize(bounds)
+            # don't use the `output_size` (`size_at_requested`) here
+            # because the rounding error at `bounds_at_baseline` leads to
+            # different `size_at_requested` (keeping same read resolution
+            # but base image is of different scale)
+            (
+                read_level,
+                bounds_at_read_level,
+                _,
+                post_read_scale,
+            ) = self._find_read_bounds_params(
+                bounds_at_baseline, resolution=resolution, units=units
+            )
+        else:  # duplicated portion with VirtualReader, factoring out ?
+            # Find parameters for optimal read
+            (
+                read_level,
+                bounds_at_read_level,
+                size_at_requested,
+                post_read_scale,
+            ) = self._find_read_bounds_params(
+                bounds_at_baseline, resolution=resolution, units=units
+            )
+
+        wsi = self.wsi
+
+        # Read at optimal level and corrected read size
+        location_at_baseline = bounds_at_baseline[:2]
+        level_location, size_at_read_level = utils.transforms.bounds2locsize(
+            bounds_at_read_level
+        )
+        level_size = self.info.level_dimensions[read_level]
+        read_bounds = utils.image.find_overlap(
+            level_location, size_at_read_level, level_size
+        )
+        _, read_size = utils.transforms.bounds2locsize(read_bounds)
+        im_region = wsi.read_region(
```

Comment:
not sure how you are reading a region as read_region is supposed to use read_rect.